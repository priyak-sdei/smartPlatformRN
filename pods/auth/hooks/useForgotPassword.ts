import { useState } from 'react';

interface UseForgotPassword {
  email: string;
  setEmail: (val: string) => void;
  isLoading?: boolean;
  error?: string | null;
  successMessage?: string | null;
  handleForgotPassword: () => Promise<void>;
  validateForm?: () => boolean;
  isValid: () => boolean;
}

// Allow injection of async logic for testability
export const useForgotPassword = (
  requestPasswordReset?: (email: string) => Promise<void>
): UseForgotPassword => {
  const [email, setEmail] = useState('');

  const validateForm = () => {
    if (!email.trim() || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)) {
      return false;
    }
    return true;
  };

  const defaultRequest = async (email: string) => {
    await new Promise(resolve => setTimeout(resolve, 1500));
    console.log('Simulated password reset request successful for:', email);
  };

  const handleForgotPassword = async () => {
    if (!validateForm()) {
      return;
    }

    try {
      await (requestPasswordReset || defaultRequest)(email);
      setEmail(''); // Optionally clear the email field on success
    } catch (err: any) {
      console.error('Forgot password error:', err);
    } finally {
    }
  };
  const isValid = () => validateForm();

  return {
    email,
    setEmail,
    handleForgotPassword,
    isValid,
  };
};
