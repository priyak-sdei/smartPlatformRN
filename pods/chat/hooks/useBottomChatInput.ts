import { pickImagesFromGallery } from '@src/helpers/mediaPicker';
import { useState } from 'react';
import { Platform } from 'react-native';

// Bottom chat input hooks
export const useBottomChatInput = (
  onSendMessage: (message: string) => void,
  maxLength: number,
) => {
  const [inputMessage, setInputMessage] = useState<string>('');
  // Function to handle sending messages
  const handleSend = (): void => {
    const trimmedMessage = inputMessage.trim();
    if (trimmedMessage.length > 0) {
      onSendMessage(trimmedMessage);
      setInputMessage('');
    }
  };

  // Function to handle text input changes
  const handleTextChange = (text: string): void => {
    if (text.length <= maxLength) {
      setInputMessage(text);
    }
  };

  const trimmedLength = inputMessage.trim().length;
  const isSendDisabled = trimmedLength === 0 || trimmedLength > maxLength;

  // Function to handle attaching images
  const handleAttach = async () => {
    const images = await pickImagesFromGallery();

    if (images?.success && images.data) {
      console.log('images', images);
      let imagePath;
      if (Array.isArray(images.data)) {
        imagePath =
          Platform.OS === 'ios'
            ? images.data[0]?.sourceURL
            : images.data[0]?.path;
      } else {
        imagePath =
          Platform.OS === 'ios' ? images.data?.sourceURL : images.data?.path;
      }

      if (imagePath) {
        onSendMessage(imagePath); // Send image path
      } else {
        // Optionally, handle the case where no image is selected (e.g., show a message or do nothing)
        // For now, do nothing if no image path is found
      }
    }
  };
  return {
    inputMessage,
    handleSend,
    handleTextChange,
    handleAttach,
    isSendDisabled,
  };
};
