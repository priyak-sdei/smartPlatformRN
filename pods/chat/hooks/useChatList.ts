import { ChatNavigationProp } from '../types/navigation';
import { ListItem } from '../types/components';
import { useState } from 'react';

/**
 * Custom hook to handle chat list functionality
 *
 * @param navigation - Navigation prop for handling screen navigation
 * @returns Object containing methods for chat list interactions
 */
export const useChatList = (
  navigation: ChatNavigationProp<'ChatList'>,
  data: ListItem[],
) => {
  const [listMessages, setListMessages] = useState<ListItem[]>(data);
  /**
   * Handles press event on a chat list item
   * Navigates to the conversation screen with the selected item data
   *
   * @param item - The chat list item that was pressed
   */
  const handleItemPress = (item: ListItem) => {
    navigation.navigate('ConversationScreen', { item: item });
  };

  const handleChangeText = (text: string) => {
    console.log('text', text);
    const filteredData = data.filter(item =>
      item.name.toLowerCase().includes(text.toLowerCase()),
    );
    setListMessages(filteredData);
    if (text === '') {
      setListMessages(data);
    }
  };

  return {
    handleItemPress,
    handleChangeText,
    listMessages,
  };
};
