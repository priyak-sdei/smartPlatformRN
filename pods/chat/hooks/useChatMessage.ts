import { useEffect, useRef } from 'react';
import { FlatList, Keyboard, Platform, KeyboardEvent } from 'react-native';
import { Message } from '../types/components';

/**
 * Props interface for useChatMessages hook
 * @interface UseChatMessagesProps
 * @property {Message[]} messages - Array of chat messages to display
 * @property {boolean} [autoScroll=true] - Whether to auto-scroll to bottom on new messages
 */
interface UseChatMessagesProps {
  messages: Message[];
  autoScroll?: boolean;
}

/**
 * Custom hook to manage chat messages scrolling behavior
 * Handles auto-scrolling on new messages and keyboard show events
 *
 * @param {UseChatMessagesProps} props - Hook configuration props
 * @returns {Object} Object containing flatListRef for the chat messages list
 */
export const useChatMessages = ({
  messages,
  autoScroll = true,
}: UseChatMessagesProps) => {
  // Reference to the FlatList component
  const flatListRef = useRef<FlatList>(null);

  /**
   * Scrolls the chat to the bottom with animation
   * Uses a small delay to ensure proper scroll behavior
   */
  const scrollToBottom = () => {
    setTimeout(() => {
      flatListRef.current?.scrollToEnd({ animated: true });
    }, 100);
  };

  // Auto-scroll when new messages are added
  useEffect(() => {
    if (autoScroll && messages.length > 0) {
      scrollToBottom();
    }
  }, [messages.length, autoScroll]);

  // Handle keyboard appearance
  useEffect(() => {
    const keyboardDidShow = Keyboard.addListener(
      Platform.OS === 'ios' ? 'keyboardWillShow' : 'keyboardDidShow',
      (_e: KeyboardEvent) => {
        scrollToBottom();
      },
    );

    // Cleanup listener on component unmount
    return () => {
      keyboardDidShow.remove();
    };
  }, []);

  return {
    flatListRef,
  };
};
