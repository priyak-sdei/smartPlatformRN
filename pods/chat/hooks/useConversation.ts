import { useState } from 'react';

import { messageData } from '../data';
import { Message } from '../types/components';

export interface User {
  _id: string;
  name: string;
  avatar: string;
}

export interface ConversationItem {
  id?: string;
  [key: string]: any;
}

export const useConversation = (_item: ConversationItem) => {
  // Replace with your actual current user ID
  const currentUserId = 'current-user-id';

  // Mock messages data - replace with your actual messages state management
  const [messages, setMessages] = useState<Message[]>(messageData);

  const handleSendMessage = (messageText: string): void => {
    const newMessage: Message = {
      _id: Date.now().toString(),
      text: messageText,
      createdAt: new Date(),
      user: {
        _id: currentUserId,
        name: 'You',
        avatar: 'https://randomuser.me/api/portraits/women/1.jpg',
      },
    };

    setMessages(prevMessages => [...prevMessages, newMessage]);
  };

  return {
    messages,
    currentUserId,
    handleSendMessage,
  };
};
