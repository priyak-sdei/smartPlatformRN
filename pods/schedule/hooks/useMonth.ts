import moment from 'moment';
import { useState } from 'react';

export function useMonth() {
  const [selectedDay, setSelectedDay] = useState(
    moment(new Date()).format('YYYY-MM-DD'),
  );
  const [markedDates, setMarkedDates] = useState({});
  const onDayPress = (day: any) => {
    setSelectedDay(day.dateString);
    setMarkedDates(day.dateString);
  };
  return {
    selectedDay,
    markedDates,
    onDayPress,
  };
}
