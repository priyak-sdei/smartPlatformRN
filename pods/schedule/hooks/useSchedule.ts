import { useState } from 'react';
import { Mode } from 'react-native-big-calendar';

export function useSchedule() {
  const events = [
    {
      title: 'Meeting',
      start: new Date(2025, 1, 11, 10, 0),
      end: new Date(2020, 1, 11, 10, 30),
    },
    {
      title: 'Coffee break',
      start: new Date(2025, 1, 11, 15, 45),
      end: new Date(2020, 1, 11, 16, 30),
    },
  ];
  const [calendarMode, setCalendarMode] = useState<Mode | undefined>('month');

  return {
    events,
    calendarMode,

    setCalendarMode,
  };
}
