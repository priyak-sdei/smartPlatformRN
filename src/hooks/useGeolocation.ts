import Geolocation from '@react-native-community/geolocation';
import { requestLocationPermission } from '@src/helpers/permission';
import {
  GeoPosition,
  GeolocationError,
  GeolocationState,
  Location,
} from '@src/types/GeoLocation';
import { useRef, useState } from 'react';
import { AnimatedRegion } from 'react-native-maps';

export const useGeolocation = () => {
  const initialLocation: Location = {
    latitude: 0,
    longitude: 0,
  };

  const [state, setState] = useState<GeolocationState>({
    currLocation: initialLocation,
    destination: null,
    isLoading: false,
    coordinate: new AnimatedRegion({
      latitude: initialLocation.latitude,
      longitude: initialLocation.longitude,
      latitudeDelta: 0.01,
      longitudeDelta: 0.01,
    }),
    time: 0,
    distance: 0,
  });

  const { currLocation, destination, coordinate, time, distance } = state;

  const updateState = (data: Partial<GeolocationState>) =>
    setState(prev => ({ ...prev, ...data }));

  /**
   * Updates the loading state
   * @param isLoading - The loading state to set
   */
  const setLoading = (isLoading: boolean) => {
    updateState({ isLoading });
  };

  /**
   * Gets the current location of the device
   * @returns Promise<void>
   */
  const getCurrentLocation = async () => {
    try {
      setLoading(true);
      const granted = await requestLocationPermission();
      if (!granted) {
        console.warn('Location permission not granted');
        return;
      }
      Geolocation.getCurrentPosition(
        (position: GeoPosition) => {
          const { latitude, longitude } = position.coords;
          const newLocation = { latitude, longitude };
          updateState({ currLocation: newLocation });
          console.log('newLocation', newLocation);
        },
        (error: GeolocationError) => {
          console.error('Geolocation error:', error);
        },
        {
          enableHighAccuracy: true,
          timeout: 15000,
          maximumAge: 10000,
        },
      );
    } catch (error) {
      console.error('Error getting current location:', error);
    } finally {
      setLoading(false);
    }
  };

  const watchIdRef = useRef<number | null>(null);

  /**
   * Starts watching the device's position
   */
  const startWatchPosition = () => {
    try {
      const watchId = Geolocation.watchPosition(
        (position: GeoPosition) => {
          const { latitude, longitude } = position.coords;
          const newCoordinate = { latitude, longitude };

          updateState({ currLocation: newCoordinate });

          // Animate the coordinate change
          // coordinate
          //   .timing({
          //     toValue: coordinate.create({
          //       latitude: newCoordinate.latitude,
          //       longitude: newCoordinate.longitude,
          //       latitudeDelta: 0.01,
          //       longitudeDelta: 0.01,
          //     }),
          //     useNativeDriver: false,
          //     latitude: 0.01,
          //     longitude: 0.01,
          //     latitudeDelta: 0.01,
          //     longitudeDelta: 0.01,
          //   })
          //   .start();
        },
        (error: GeolocationError) => {
          console.error('Watch position error:', error);
        },
        {
          enableHighAccuracy: true,
          distanceFilter: 10,
          interval: 5000,
          fastestInterval: 2000,
        },
      );

      watchIdRef.current = watchId;
    } catch (error) {
      console.error('Error starting watch position:', error);
    }
  };

  const stopWatchPosition = () => {
    if (watchIdRef.current !== null) {
      Geolocation.clearWatch(watchIdRef.current);
      watchIdRef.current = null;
    }
  };

  /**
   * Updates the distance and time values
   * @param distance - The distance value to set
   * @param time - The time value to set
   */
  const fetchTime = () => {
    updateState({
      distance,
      time,
    });
  };

  /**
   * Sets the destination location
   * @param location - The destination location to set
   */
  const setDestination = (location: Location) => {
    updateState({ destination: location });
  };

  /**
   * Cleanup function to be called when component unmounts
   */
  const cleanup = () => {
    stopWatchPosition();
  };

  return {
    currLocation,
    destination,
    coordinate,
    time,
    distance,
    isLoading: state.isLoading,
    getCurrentLocation,
    startWatchPosition,
    stopWatchPosition,
    fetchTime,
    setDestination,
    cleanup,
  };
};
