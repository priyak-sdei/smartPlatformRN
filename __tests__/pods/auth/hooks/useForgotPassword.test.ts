import { act, renderHook } from '@testing-library/react-hooks';
import { useForgotPassword } from '../../../../pods/auth/hooks/useForgotPassword';

describe('useForgotPassword', () => {
    it('should initialize with empty email', () => {
        const { result } = renderHook(() => useForgotPassword());
        expect(result.current.email).toBe('');
        expect(result.current.isValid()).toBe(false);
    });

    it('should update email', () => {
        const { result } = renderHook(() => useForgotPassword());
        act(() => {
            result.current.setEmail('test@example.com');
        });
        expect(result.current.email).toBe('test@example.com');
    });

    it('should validate correct email', () => {
        const { result } = renderHook(() => useForgotPassword());
        act(() => {
            result.current.setEmail('test@example.com');
        });
        expect(result.current.isValid()).toBe(true);
    });

    it('should invalidate incorrect email', () => {
        const { result } = renderHook(() => useForgotPassword());
        act(() => {
            result.current.setEmail('invalid-email');
        });
        expect(result.current.isValid()).toBe(false);
    });

    it('should not call API if email is invalid', async () => {
        const { result } = renderHook(() => useForgotPassword());
        act(() => {
            result.current.setEmail('bad');
        });
        // handleForgotPassword should return early
        await act(async () => {
            await result.current.handleForgotPassword();
        });
        // Email should remain unchanged
        expect(result.current.email).toBe('bad');
    });

    it('should simulate API and clear email on success', async () => {
        const { result } = renderHook(() => useForgotPassword());
        act(() => {
            result.current.setEmail('test@example.com');
        });
        await act(async () => {
            await result.current.handleForgotPassword();
        });
        expect(result.current.email).toBe('');
    });

    it('should handle errors in handleForgotPassword', async () => {
        const errorFn = jest.fn(() => Promise.reject(new Error('Simulated error')));
        const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => { });
        const { result } = renderHook(() => useForgotPassword(errorFn));
        act(() => {
            result.current.setEmail('test@example.com');
        });
        await act(async () => {
            await result.current.handleForgotPassword();
        });
        expect(errorFn).toHaveBeenCalledWith('test@example.com');
        errorSpy.mockRestore();
    });

    it('matches snapshot', () => {
        const { result } = renderHook(() => useForgotPassword());
        expect(result.current).toMatchSnapshot('useForgotPassword hook');
    });
}); 