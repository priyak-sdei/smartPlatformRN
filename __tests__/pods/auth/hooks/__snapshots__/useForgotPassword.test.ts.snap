// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForgotPassword matches snapshot: useForgotPassword hook 1`] = `
{
  "email": "",
  "handleForgotPassword": [Function],
  "isValid": [Function],
  "setEmail": [Function],
}
`;
